src/cpp/rtps/writer/RTPSWriter.cpp

```diff
@@ -18,40 +18,111 @@
  */
 
 #include <fastdds/rtps/writer/RTPSWriter.h>
+
+#include <fastdds/dds/log/Log.hpp>
+
 #include <fastdds/rtps/history/WriterHistory.h>
 #include <fastdds/rtps/messages/RTPSMessageCreator.h>
-#include <fastdds/dds/log/Log.hpp>
-#include <rtps/participant/RTPSParticipantImpl.h>
+
+#include <rtps/history/BasicPayloadPool.hpp>
+#include <rtps/history/CacheChangePool.h>
 #include <rtps/flowcontrol/FlowController.h>
+#include <rtps/participant/RTPSParticipantImpl.h>
 
 #include <mutex>
 
 namespace eprosima {
 namespace fastrtps {
 namespace rtps {
 
+static std::shared_ptr<IChangePool> create_change_pool(
+        const std::shared_ptr<IPayloadPool>& payload_pool,
+        const PoolConfig& pool_config)
+{
+    if ((pool_config.memory_policy == PREALLOCATED_MEMORY_MODE) ||
+            (pool_config.memory_policy == PREALLOCATED_WITH_REALLOC_MEMORY_MODE))
+    {
+        auto init_cache = [payload_pool, pool_config](
+            CacheChange_t* item)
+                {
+                    if (payload_pool->get_payload(pool_config.payload_initial_size, *item))
+                    {
+                        payload_pool->release_payload(*item);
+                    }
+                };
+
+        return std::make_shared<CacheChangePool>(pool_config, init_cache);
+    }
+
+    return std::make_shared<CacheChangePool>(pool_config);
+}
```

Comment:
We could put this implementation in `RTPSWriter::create_change_pool` and avoid an extra function.

Reply:
I refactored this with a new method on BasicPayloadPool returning both pools.